packages/@aws-cdk/aws-cognito/lib/identity-pool.ts

```diff
@@ -0,0 +1,729 @@
+    return client;
+  }
+
+  /**
+   * Adds Role Mappings to Identity Pool
+  */
+  public addRoleMappings(...mappings: IdentityPoolRoleMapping[]): void {
+    if (!mappings || !mappings.length) return;
+    const name = this.id + 'RoleMappingAttachment';
+    const roleMappings = this.configureRoleMappings(...mappings);
+
+    const attachment = new CfnIdentityPoolRoleAttachment(this, name, {
+      identityPoolId: this.identityPoolId,
+      roles: {},
+      roleMappings,
+    });
+    attachment.node.addDependency(this.cfnIdentityPool);
+  }
+
+  /**
+   * Add permissions to default role for authenticated users
+   */
+  public addUserPermissions(...statements: PolicyStatementProps[]): void {
+    statements.forEach(statement => this.authenticatedRole.addToPolicy(
+      new PolicyStatement(statement),
+    ));
+  }
+
+  /**
+   * Add permissions to default role for unauthenticated users
+   */
+  public addGuestPermissions(...statements: PolicyStatementProps[]): void {
+    statements.forEach(statement => this.unauthenticatedRole.addToPolicy(
+      new PolicyStatement(statement),
+    ));
+    // Make sure `allowUnauthenticatedIdentities` is true now that there are guest permissions
+    (this.node.defaultChild as CfnIdentityPool).addPropertyOverride('AllowUnauthenticatedIdentities', true);
+  }
+
+  /**
+   * Configure Default Roles For Identity Pool
+  */
+  private configureDefaultRole(
+    type: string,
+    additionalPermissions: PolicyStatementProps[] = [],
+    action: string = 'sts:AssumeRoleWithWebIdentity',
+  ): Role {
+    const name = this.id + type + 'Role';
+    const role = new Role(this, name, {
+      roleName: name,
+      description: `Default ${type} Role for Identity Pool ${this.identityPoolName}`,
+      assumedBy: new FederatedPrincipal('cognito-identity.amazonaws.com', {
+        'StringEquals': {
+          'cognito-identity.amazonaws.com:aud': this.identityPoolId,
+        },
+        'ForAnyValue:StringLike': {
+          'cognito-identity.amazonaws.com:amr': type.toLowerCase(),
+        },
+      }, action),
+    });
+    additionalPermissions.forEach(statement => role.addToPolicy(
+      new PolicyStatement(statement),
+    ));
+    return role;
+  }
+
+  /**
+   * Configure Role Attachments For Identity Pool
+  */
+  private configureRoleAttachment(
+    authenticatedRole?: IRole,
+    unauthenticatedRole?: IRole,
+    ...mappings: IdentityPoolRoleMapping[]
+  ): void {
+    const name = this.id + 'RoleAttachment';
+    let roles: any = undefined, roleMappings: any = undefined;
+    if (authenticatedRole || unauthenticatedRole) {
+      roles = {};
+      if (authenticatedRole) roles.authenticated = authenticatedRole.roleArn;
+      if (unauthenticatedRole) roles.unauthenticated = unauthenticatedRole.roleArn;
+    }
+    if (mappings) {
+      roleMappings = this.configureRoleMappings(...mappings);
+    }
+    const attachment = new CfnIdentityPoolRoleAttachment(this, name, {
```

Comment:
I struggled with this one, and in the end I couldn't think of a use case where an `IdentityPoolRoleAttachment` would be used independently of an `IdentityPool`. And also an `IdentityPool` would be useless without at least one corresponding `IdentityPoolRoleAttachment`. They have to go hand in hand. So it seemed that the simplest way to implement role attachments was to effectively combine them with `IdentityPool`.

If there's a policy to make an L2 Construct for every L1 Construct, I would suggest making the `IdentityPoolRoleAttachment` a private construct. I don't see a need for it to ever be used directly. What do you think?

Reply:
Ok. in that case, I think it's fine to leave this internal as you already have it here.

Reply:
Ok well I ended up adding it because given the changes made to the `IdentityPool` Construct there are a handful of use cases where the `IdentityPoolRoleAttachment` can be used now, even though for the most part it's still under the hood. But I think I got it to a much better place, so tell me what you think.